### ISSUE TITLE
SslStream AuthenticateAsClient method is not sending SNI information to the server on OS X

### ISSUE THREAD

**Issue by rstam** [dotnet/corefx#17427]
When using AuthenticateAsClient to connect to a server, SslStream does not appear to be sending the SNI information to the server.

We have observed this issue on OS X. When running on Windows the SNI information *does* appear to be sent. It is unknown to us whether the SNI information is sent when running on Linux.

**Comment by davidsh** [comment]
Not sure if this is related to #9608.  SslStream has limited supported for SNI in general and does not support SNI on the server-side (`AuthenticateAsServer`).

**Comment by mongostephen** [comment]
Just a note, on Windows, SslStream appears to pass on SNI.

**Comment by Priya91** [comment]
The client Sslstream on Linux also sends sni headers.

**Comment by karelz** [comment]
Seems to have impact on MongoDB client driver, moving to 2.1

**Comment by Priya91** [comment]
I wrote a small app on MacOS and hit a websocket server using ManagedClientWebSocket and an https server using ManagedHttpClientHandler, both of these types uses SslStream in their implementation. I verified the ssl handshake data sent over the wire using wireshark, and verified that it in fact sends the server_name TLS extension in the ClientHello. Can you provide a small repro app for this bug, along with the ssl handshake header information?

**Comment by karelz** [comment]
Closing, feel free to reopen when there is a repro we can look at.

**Comment by ar7z1** [comment]
@karelz @Priya91 I've faced the same problem on my linux box. I use `dotnet-runtime-2.0.5`.

A small program to reproduce the issue:
```csharp
using System;
using System.Net.Security;
using System.Net.Sockets;
using System.Security.Authentication;

namespace TestSNI
{
    class Program
    {
        static void Main(string[] args)
        {
            var host = "cluster0-shard-00-00-fvaks.mongodb.net";
            using (var client = new TcpClient(host, 27017))
            {
                using (var sslStream = new SslStream(client.GetStream(), false))
                {
                    sslStream.AuthenticateAsClient(host);
                }
            }
        }
    }
}
```

On Windows this code sends SNI extension:
![image](https://user-images.githubusercontent.com/2307197/35973659-54a8edf0-0ce7-11e8-9c62-dd9ede84d014.png)

But on Linux (I've tested on Ubuntu 14.04 and Ubuntu 16.04) it doesn't:
![image](https://user-images.githubusercontent.com/2307197/35973736-93583f92-0ce7-11e8-837a-d258a9bcfb8b.png)

Pcaps for Windows, Ubuntu 14.04 and Ubuntu 16.04: [SNI.zip](https://github.com/dotnet/corefx/files/1707028/SNI.zip).

I think that the problem is because awesome fix from @Priya91 (#25118) doesn't exist in v2.0.5: https://github.com/dotnet/corefx/blob/v2.0.5/src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs#L108

**Comment by karelz** [comment]
Yes, client-side SNI was added into 2.1. Did you try to run it on 2.1? (see [dogfooding](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md))

**Comment by ar7z1** [comment]
@karelz I've checked the master nightly build, everything is perfectly working!
Is there any chance to release this feature before 2.1?

**Comment by ar7z1** [comment]
> the 2.0 servicing fix might be available quite close to 2.1 availability

@karelz Then I'll wait till 2.1. :-)

Thank you again for awesome work!
